### PR TITLE
[Fix #9144] Add `aggressive` and `conservative` enforced styles for `Style/StringConcatenation` cop

### LINKED ISSUE #9144: Style/StringConcatenation mistakenly assumes that the + operator means concatenation, e.g. with Pathname
## Expected behavior

Rubocop should not suggest replacing an expression that returns a `Pathname` with an expression that returns a `String`, even if we use the `+` operator to combine a `Pathname` and `String` (which [returns a `Pathname`](https://ruby-doc.org/stdlib-2.6.3/libdoc/pathname/rdoc/Pathname.html#method-i-2B))

## Actual behavior

```
1.5.0 (using Parser 2.7.2.0, rubocop-ast 1.3.0, running on ruby 2.6.6 x86_64-linux)
Inspecting 1 file
C

Offenses:

bad.rb:4:7: C: Style/StringConcatenation: Prefer string interpolation to string concatenation.
puts (Pathname.new('/') + 'test').class
      ^^^^^^^^^^^^^^^^^^^^^^^^^^
```

## Steps to reproduce the problem

Here are the files I used to produce the error:

`Gemfile`
```ruby
source 'https://rubygems.org'
gem 'rubocop', '=1.5.0'
```

`bad.rb`
```ruby
# frozen_string_literal: true

require 'pathname'
puts (Pathname.new('/') + 'test').class
```

> Note that running the `bad.rb` will print `Pathname`

### Faster way to reproduce the issue

```Dockerfile
FROM ruby:2.6-slim
ARG TESTFILE=bad.rb

RUN true \
  && echo "source 'https://rubygems.org'" > Gemfile \
  && echo "gem 'rubocop', '=1.5.0'" >> Gemfile \
  && cat Gemfile \
  && bundle install

RUN true \
  && echo "# frozen_string_literal: true\n" > $TESTFILE \
  && echo "require 'pathname'" >> $TESTFILE \
  && echo "puts (Pathname.new('/') + 'test').class" >> $TESTFILE \
  && cat $TESTFILE

RUN ruby $TESTFILE
RUN bundle exec rubocop -V && bundle exec rubocop --debug -D $TESTFILE
```

Saving this as `Dockerfile` and running `docker build .` should show the entire process

### PR DESCRIPTION
**Pull request by tejasbubane**
Closes #9144

-----------------

Before submitting the PR make sure the following are checked:

* [x] The PR relates to *only* one subject with a clear title and description in grammatically correct, complete sentences.
* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Ran `bundle exec rake default`. It executes all tests and runs RuboCop on its own code.
* [x] Added an entry (file) to the [changelog folder](https://github.com/rubocop-hq/rubocop/blob/master/changelog/) named `{change_type}_{change_description}.md` if the new code introduces user-observable changes. See [changelog entry format](https://github.com/rubocop-hq/rubocop/blob/master/CONTRIBUTING.md#changelog-entry-format) for details.

[1]: https://chris.beams.io/posts/git-commit/

### CODE REVIEW

**Comment by ianfixes** [comment]
This is a small step forward, but I want to be clear that my aim in opening #9144 was simply to show that the _+ operator_ could not be trusted to indicate a concatenation operation.  What's being addressed in this PR is only my very contrived counterexample.  

A more common counterexample would be in a form like this:

```ruby
def for_example(a_pathname)
    a_pathname + "a_string"
end
```

RuboCop really has no business pretending that string concatenation is the _only_ thing that `+ "a_string"` could be accomplishing.

**Comment by bbatsov** [comment]
I also don't like the proposed change as I don't think the problem is solvable by ignoring some cases. Fundamentally, the problem is that with operator overloading we can never know what happens in certain cases and we can only guess. The cop is already marked as unsafe because of this. The only thing I can think of is to add a mode that ignores cases where the first operand is not a string literal.

**Comment by tejasbubane** [comment]
> This is a small step forward, but I want to be clear that my aim in opening #9144 was simply to show that the _+ operator_ could not be trusted to indicate a concatenation operation. What's being addressed in this PR is only my very contrived counterexample.

Yes I had the same thing in mind, but thought `Pathname` being stdlib made sense to add an exception.

> The only thing I can think of is to add a mode that ignores cases where the first operand is not a string literal.

That sounds a lot better. I will change this PR accordingly.

**Comment by ianfixes** [comment]
> The only thing I can think of is to add a mode that ignores cases where the first operand is not a string literal.

Cross posting my latest comment from #9144, I think this could be expanded to say "only consider cases where _the first and last operands_ in a chain of `+` operators are string lterals"

```ruby
def some_fn(var_of_unknown_type)
  puts "hello" + "world"                       # string concatenation 
  puts var_of_unknown_type + "world"           # insufficient info to decide
  puts "hello" + var_of_unknown_type           # string concatenation (I think? Is there a case where that's not true?)
  puts "hello" + var_of_unknown_type + "world" # string concatenation 
end
```

**Comment by tejasbubane** [comment]
@bbatsov Changed to use a config option.

File: lib/rubocop/cop/style/string_concatenation.rb [hunk 1]
```diff
@@ -11,16 +11,24 @@ module Style
       # In those cases, it might be useful to extract statements to local
       # variables or methods which you can then interpolate in a string.
       #
-      # @example
+      # @example EnforcedStyle: all (default)
```
Comment (bbatsov) [inline lib/rubocop/cop/style/string_concatenation.rb:14]: I'd expand the cop docs to explain the safety issues and the implications of the using the different styles.

**Comment by bbatsov** [comment]
I'm also thinking we can have some more generic names for the enforced style (e.g. `conservative` and `aggressive`) or perhaps it'd be even better if this was some boolean flag like `SafeMode`. We're not really supporting different styles here, we're rather limitting the scope of the cop.

**Comment by bbatsov** [comment]
Ping :-)

**Comment by tejasbubane** [comment]
I forgot about this PR. I will fix this in a day or two.

**Comment by bbatsov** [comment]
@tejasbubane another friendly reminder about this PR :-)

**Comment by tejasbubane** [comment]
@bbatsov Changes done. Sorry for the delay.

File: lib/rubocop/cop/style/string_concatenation.rb [hunk 2]
```diff
@@ -15,7 +15,13 @@ module Style
       # lines, this cop does not register an offense; instead,
       # `Style/LineEndConcatenation` will pick up the offense if enabled.
       #
-      # @example
+      # Two modes are supported:
```
Comment (bbatsov) [inline lib/rubocop/cop/style/string_concatenation.rb:18]: You should also mention how this ties to the cop safety - e.g. with the `Pathname` example.

**Comment by bbatsov** [comment]
Seems I totally forgot about this myself. After some further thought let's not use `EnforcedStyle`, but rather some other configuration option name (e.g. `Mode/Detection`), as that's not really a different style just an instruction how hard to try to find offenses. If someone from @rubocop/rubocop-core can suggest a better config setup - be my guest.

**Review by bbatsov — CHANGES_REQUESTED** [review]
Overall things look good. We just need to add some safety documentation and adopt some dedicated configuration option for the two modes of operation.

File: spec/rubocop/cop/style/string_concatenation_spec.rb [hunk 3]
```diff
@@ -201,4 +201,33 @@
...
   end
+
+  context 'EnforcedStyle = conservative' do
+    let(:cop_config) { { 'EnforcedStyle' => 'conservative' } }
+
+    context 'when first operand is not string literal' do
+      it 'does not register offense' do
+        expect_no_offenses(<<~RUBY)
+          user.name + "!!"
+          user.name + "<" + "user.email" + ">"
```
Comment (bquorning) [inline spec/rubocop/cop/style/string_concatenation_spec.rb:212]: Actually, I would expect an offense here. Not on `user.name + "<"`, but on `"<" + "user.email"` (and `"user.email" + ">"`).

Comment (bbatsov) [reply spec/rubocop/cop/style/string_concatenation_spec.rb:212]: Good catch!

Comment (ianfixes) [reply spec/rubocop/cop/style/string_concatenation_spec.rb:214]: I disagree.

```ruby
# yes, I know that a variable called user.name is unlikely to be a Pathname
user.name = Pathname.new("user.name")
result = user.name + "<" + "user.email" + ">"
puts "result is #{result}" # result is user.name/</user.email/>
                           # which is in no way equivalent to "user.name<user.email>"
```

**Comment by tejasbubane** [comment]
Review comments fixed. 

* Changed config name to `Mode`. 
* Fixed the issue @bquorning pointed out. 
* Added `Pathname` example to cop documentation to show how the config ties to cop safety.

File: spec/rubocop/cop/style/string_concatenation_spec.rb (hunk 3 above)
Comment (ianfixes) [reply spec/rubocop/cop/style/string_concatenation_spec.rb]: @tejasbubane -- I think the fix that was called for here was in error.  See my above comment

File: lib/rubocop/cop/style/string_concatenation.rb [hunk 4]
```diff
@@ -15,18 +15,37 @@ module Style
...
+      # 1. `aggressive` style checks and corrects all occurrences of `+` where
+      # either the left or right side of `+` is a string literal.
+      # 2. `conservative` style on the other hand, checks and corrects only if
+      # left side (receiver of `+` method call) is a string literal.
+      # This is useful when the receiver is some expression that returns string like `Pathname`
```
Comment (ianfixes) [inline lib/rubocop/cop/style/string_concatenation.rb:23]: `Pathname.new` returns a `Pathname`, not a string.  `Pathname.new("x") + "y"` also returns a `Pathname`.

I think this comment should read 
> This is useful when the receiver is an object whose `+` operator accepts a string, but does not return a string -- such as `Pathname`

File: changelog/change_string_concatenation_enforced_styles.md [hunk 5]
```diff
@@ -0,0 +1 @@
+* [#9144](https://github.com/rubocop/rubocop/issues/9144): Add `aggressive` and `conservative` enforced styles for `Style/StringConcatenation` cop. ([@tejasbubane][])
```
Comment (bbatsov) [inline changelog/change_string_concatenation_enforced_styles.md:1]: Those are not enforced styles, but "modes of operation".

File: spec/rubocop/cop/style/string_concatenation_spec.rb [hunk 6]
```diff
@@ -203,4 +203,36 @@
       RUBY
     end
   end
+
+  context 'EnforcedStyle = conservative' do
```
Comment (bbatsov) [inline spec/rubocop/cop/style/string_concatenation_spec.rb:207]: Mode

**Comment by bbatsov** [comment]
@tejasbubane ping :-)

**Comment by tejasbubane** [comment]
@bbatsov Minor wording changes done.

**Review by bbatsov — APPROVED** [review]

**Pull request merged**